README.md
# Swusana

Asana is great but some bits are annoying.  The list view is fixed at 1000px.  There is a lot of noise on the detail page with stories telling you about who moved a task where and these sometimes obsuce the actual human comments.
If you run meetings and take notes into Asana you get added as a follower to every task which then pollutes your inbox with many notifications.

This is a Tamper Monkey script to add buttons to the top nav bar of Asana that let you togle the following functionality: 
- Noise reduction and beautification: hide noise in the comments.
- Auto-unfollow: A blackout period button that auto-unfollows you from all tasks while it is activated.
- Image: Set a custom background image from a URL.
- Timer: Show a counter for a selected ticket (list view only) 

Currently only tested in Chrome.

## Installation

- Install tampermonkey: https://chrome.google.com/webstore/detail/tampermonkey/dhdgffkkebhmkfjojejmpbldmpobfkfo?hl=en
- Click on the below raw script URL that should auto install the script into Tamper Monkey: https://github.com/sendwithus/swusana/raw/master/swusana.user.js
- Reload your Asana page and you should see new buttons appear in the top nav bar.


changelog.md
# Changelog

0.14.0
--
- BUGFIX: bug/feature templates selector updated to reflect change in Asana source

0.13.0
--
- updated a selector to reflect change in Asana source

0.12.0
--
- Alphabetize the custom fields.

0.11.2
--
- Fixed title on add feature button.

0.11.1
--
- Fixed bug where blackout icon changes to the wrong icon when unselected.

0.11.0
--
- Added two buttons to create bug / feature tasks with a template.  The template is loaded from the the swusana github directory /templates.

0.10.3
--
- Remove the upgrade button from the top bar.

0.10.0
--
- Timer that changes colour as you spend time on a ticket.  Resets when a new ticket is selected.

0.9.1
--
- Turn off image when not used.

0.9.0
--
- Can now set a background image and the opacity is set to 0.93 for the foreground boxes.

0.8.9
--
- Asana breaking style change fix 

0.8.2
-- 
- Red glowing dot when you turn on auto-unfollow

0.8.1
--
- fixed auto-unfollow on large screens bug

0.8.0
--
- Syntax highlighting.
 
0.6.0
--
- moved to semver.
- fixed user name linking not working.

0.5
--
- show more bug fixed.

0.4
--
- now leaves in created story when noise reduction is on.
- made `<br>` tags have correct new line markdown rendeing.

0.3
--
- Added support for blackout period.


asana-github-markdown.user.js
// ==UserScript==
// @name         Asana Github Markdown
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  Replaces comments and description with markdown interpretation.  Hides the noise.  Show it all by clicking on the description icon
// @author       will@sendwithus.com
// @match        https://app.asana.com/*
// @grant        none
// @require http://code.jquery.com/jquery-1.12.4.min.js
// @require https://cdnjs.cloudflare.com/ajax/libs/showdown/1.8.6/showdown.min.js
// ==/UserScript==
var converter = new showdown.Converter({tables: true, strikethrough: true});
function addGlobalStyle(css) {
    var head, style;
    head = document.getElementsByTagName('head')[0];
    if (!head) { return; }
    style = document.createElement('style');
    style.type = 'text/css';
    style.innerHTML = css;
    head.appendChild(style);
}
var myCss =
    '.markdowned { '+
        'white-space: normal !important; ' +
        'background-color: #FaFaFa !important;' +
        'color: #666666; '+
        'padding-left: 10px;' +
        'padding-right: 10px;' +
        'padding-top: 5px;' +
        'padding-bottom: 5px;' +
        'border-left: 4px solid #EEEEEE !important; ' +
        'display: block !important; ' +
    '}' +
    '#description-markdown { ' +
        'border-top: 1px solid #eeeeee;' +
        'border-bottom: 1px solid #eeeeee;' +
        'border-right: 1px solid #eeeeee;' +
    '}' +
    '.notagain { padding: 10px; border:1px solid #999999; }' +
    '.markdowned p { margin-top: 5px; }' +
    '.markdowned table { ' +
        'margin-top: 10px;' +
        'margin-bottom: 10px;' +
    '}' +
    '.markdowned td, .markdowned th { ' +
        'border-left: 1px solid #AAAAAA;' +
        'padding-right: 5px; ' +
        'padding-left: 4px; ' +
    '}' +
    '.markdowned td { ' +
        'border-top: 1px solid #AAAAAA;' +
    '}' +
    '.markdowned blockquote { ' +
        'border-left: 4px solid #8888FF;' +
        'padding-left:10px;' +
        'white-space: normal;' +
    '}' +
    '.markdowned blockquote p { ' +
        'margin: 0' +
    '}' +
    '.markdowned h6 { font-size:1em; font-weight: bold; margin-top: 3px; margin-bottom: 1px; }' +
    '.markdowned h5 { font-size:1.2em; margin-top: 4px; margin-bottom: 1px; }' +
    '.markdowned h4 { font-size:1.4em; margin-top: 5px; margin-bottom: 2px; }' +
    '.markdowned h3 { font-size:1.6em; margin-top: 6px; margin-bottom: 3px; }' +
    '.markdowned h2 { font-size:1.8em; margin-top: 7px; margin-bottom: 4px; }' +
    '.markdowned h1 { font-size:2em; margin-top: 8px; margin-bottom: 5px; }';

addGlobalStyle(myCss);

$(document).on('click', '.SingleTaskPane-descriptionLabel', function(){
    $('.toggleable').toggle('slide');
});


var pmInt = setInterval(function() {
    $('.RichText').not('.markdowned, .notagain').each(function(index, item){
        var md = $(item).html().replace(/<br>/g, '\n');
        var html = '<div class="markdowned">' + converter.makeHtml(md) + '</div>';
        $(item).before(html);
        $(item).addClass('notagain');
        $(item).addClass('toggleable');
        $(item).hide();
    });

    $('.SingleTaskPane-description, .StoryFeed-miniStory').not('.toggleable').each(function(index,item){
        $(item).addClass('toggleable');
        $(item).hide();
    });

    if ($('#TaskDescription-textEditor').html()) {
        var md = '';
        $('*', $('#TaskDescription-textEditor')).each(function(index, item){
            if ($(item).clone().children().remove().end().text() !== '') {
                if ($(item).prop("tagName").toLowerCase() === 'a') {
                    md += '<a href="' + $(item).text() + '">' + $(item).text() + '</a>';
                } else {
                    md += $(item).text();
                }
            }
            md += '\n';
        });
        var html = converter.makeHtml(md);
        if (!$('#description-markdown').length) {
            $('.SingleTaskPane-descriptionRow').before('<div id="description-markdown" class="markdowned RichText"></div>');
        }
        $('#description-markdown').html(html);
    }
}, 1000);


swusana.user.js
// ==UserScript==
// @name         Swusana
// @namespace    http://tampermonkey.net/
// @version      0.14.0
// @description  Asana Productivity Enhancements including - Noise Reduction.  Blackout periods.  Timer
// @author       will@sendwithus.com
// @match        https://app.asana.com/*
// @grant        GM_xmlhttpRequest
// @updateURL    https://github.com/sendwithus/swusana/blob/master/swusana.user.js
// @downloadURL  https://github.com/sendwithus/swusana/blob/master/swusana.user.js
// @require http://code.jquery.com/jquery-1.12.4.min.js
// @require https://cdnjs.cloudflare.com/ajax/libs/js-cookie/2.2.0/js.cookie.min.js
// ==/UserScript==

// TODO create some better buttons
var bugButton = $('<a id="bugButton" title="add bug" class="NavigationLink Topbar-imageButton swusana-button" href="javascript:;"><img height="24" src="https://us.123rf.com/450wm/blankstock/blankstock1409/blankstock140900675/31370510-bug-sign-icon-virus-symbol-software-bug-error-disinfection-circle-flat-button-with-shadow-modern-ui-.jpg?ver=6"></a>')
var featureButton = $('<a id="featureButton" title="add feature" class="NavigationLink Topbar-imageButton swusana-button" href="javascript:;"><img height="24" src="http://cdn.onlinewebfonts.com/svg/img_346035.png"></a>')
var imageButton = $('<a id="imageButton" title="add image" class="NavigationLink Topbar-imageButton swusana-button" href="javascript:;"><img height="24" src="https://cdn2.iconfinder.com/data/icons/media-and-navigation-buttons-round/512/Button_16-512.png"></a>')
var noiseButton = $('<a id="noiseButton" title="hide/show noise" class="NavigationLink Topbar-noiseButton swusana-button" href="javascript:;"><img height="24" src="https://icon-rainbow.com/i/icon_00671/icon_006710_256.png"></a>');
var blackoutButton = $('<a id="blackoutButton" title="turn on/off blackout period" class="NavigationLink Topbar-blackoutButton swusana-button" href="javascript:;"><img height="24" src="http://cdn.onlinewebfonts.com/svg/img_314578.png"></a>');
var timerButton = $('<a id="timerButton" title="time how long you are looking at a task" class="NavigationLink Topbar-timerButton swusana-button" href="javascript:;"><img height="24" src="https://www.freeiconspng.com/uploads/timer-icon-15.png"></a><div id="swusana-timer" style="display:none" class="swusana-timer"></div>');
var noiseButtonOn = false;
var blackoutButtonOn = false;
var imageButtonOn = false;
var timerButtonOn = false;
var blackoutProfileStyle = '';
var timer = 0;
var bugXml = '';
var featureXml = '';

GM_xmlhttpRequest({
  method: "GET",
  url: "https://raw.githubusercontent.com/sendwithus/swusana/master/templates/bug.xml",
  headers: {
    "User-Agent": "Mozilla/5.0",    // If not specified, navigator.userAgent will be used.
    "Accept": "text/xml"            // If not specified, browser defaults will be used.
  },
  onload: function(response) {
      bugXml = response.responseText;
      bugXml = bugXml.replace(/\s+/g, ' ');
      bugXml = bugXml.replace(/>\s*/g, '>');
      bugXml = bugXml.replace(/\s*</g, '<');
  }
});

GM_xmlhttpRequest({
  method: "GET",
  url: "https://raw.githubusercontent.com/sendwithus/swusana/master/templates/feature.xml",
  headers: {
    "User-Agent": "Mozilla/5.0",    // If not specified, navigator.userAgent will be used.
    "Accept": "text/xml"            // If not specified, browser defaults will be used.
  },
  onload: function(response) {
      featureXml = response.responseText;
      featureXml = featureXml.replace(/\s+/g, ' ');
      featureXml = featureXml.replace(/>\s*/g, '>');
      featureXml = featureXml.replace(/\s*</g, '<');
  }
});

// ONLOAD
var buttonBarLocation = '.TopbarHelpMenuButton';
$(document).ready(function(){
    waitForEl(buttonBarLocation, function(){
        $(buttonBarLocation).after(timerButton);
        $(buttonBarLocation).after(imageButton);
        $(buttonBarLocation).after(blackoutButton);
        $(buttonBarLocation).after(noiseButton);
        $(buttonBarLocation).after(bugButton);
        $(buttonBarLocation).after(featureButton);
        blackoutProfileStyle = $('.TopbarPageHeaderGlobalActions .Avatar').attr('style');

        $('.TopbarContingentUpgradeLink').hide();

        // TIMER TRIGGERS
        $('#timerButton').on('click', function(){
            timerButtonOn = !timerButtonOn;
            $(this).toggleClass('swusana-button-on');
            Cookies.set('timerButtonStatus', timerButtonOn, { expires: 365 });
            if (timerButtonOn) {
                $('#swusana-timer').show();
            } else {
                $('#swusana-timer').hide();
            }
        });

        if (Cookies.get('timerButtonStatus') === 'true'){
            $('#timerButton').trigger('click');
        }

        // IMAGE TRIGGERS
        $('#imageButton').on('click', function(){
            imageButtonOn = !imageButtonOn;
            $(this).toggleClass('swusana-button-on');
            Cookies.set('imageButtonStatus', imageButtonOn, { expires: 365 });
            var currentBackgroundImage = Cookies.get('currentBackgroundImage');
            if (imageButtonOn) {
                var newImage = prompt('Enter an image URL', currentBackgroundImage);
                $('#asana_ui').css('opacity', '0.93');
                $('#bg_pattern').css('background-image', 'url(' + newImage + ')').css('background-repeat','no-repeat').css('background-size','110%').addClass('themeBackground-aqua');
                Cookies.set('currentBackgroundImage', newImage, { expires: 365 });
            } else {
                $('#bg_pattern').css('background-image', '')
            }
        });

        if (Cookies.get('imageButtonStatus') === 'true'){
            $('#imageButton').toggleClass('swusana-button-on');
            var newImage = Cookies.get('currentBackgroundImage');
            $('#asana_ui').css('opacity', '0.93');
            $('#bg_pattern').css('background-image', 'url(' + newImage + ')').css('background-repeat','no-repeat').css('background-size','110%').addClass('themeBackground-aqua');;
            imageButtonOn = true;
        }

        // BLACKOUT TRIGGERS
        $('#blackoutButton').on('click', function(){
            blackoutButtonOn = !blackoutButtonOn;
            $(this).toggleClass('swusana-button-on');
            Cookies.set('blackoutButtonStatus', blackoutButtonOn, { expires: 365 });
            if (blackoutButtonOn) {
                var confirmed = confirm('Please carefully confirm.  Blackout is about to be turned on and you will be auto-unfollowed on every task you click on.\n\nClick OK to turn on.\nClick Cancel to turn off.');
                if (!confirmed){
                    blackoutButtonOn = !blackoutButtonOn;
                    Cookies.set('blackoutButtonStatus', blackoutButtonOn, { expires: 365 });
                    $(this).toggleClass('swusana-button-on');
                } else {
                    $('#blackoutButton img').attr('src', 'https://media.giphy.com/media/79VehauAiiSLS/giphy.gif');
                }
            } else {
                $('#blackoutButton img').attr('src', 'http://cdn.onlinewebfonts.com/svg/img_314578.png');
            }
        });

        if (Cookies.get('blackoutButtonStatus') === 'true'){
            $('#blackoutButton').trigger('click');
        }

        // NOISE TRIGGERS
        $('#noiseButton').on('click', function(){
            noiseButtonOn = !noiseButtonOn;
            $(this).toggleClass('swusana-button-on');
            Cookies.set('noiseButtonStatus', noiseButtonOn, { expires: 365 });
            $('.lunaui-grid-center-pane-container#center_pane_container').css({ 'max-width': '1000px !important'});
            if (!noiseButtonOn){
                $('.lunaui-grid-center-pane-container#center_pane_container').css({ 'max-width': '100% !important'});
                $('.swusana-noise').show();
                $('.swusana-noise').removeClass('swusana-noise-hidden');
            }
        });

        if (Cookies.get('noiseButtonStatus') === 'true'){
            $('#noiseButton').trigger('click');
        }

        // BUG TRIGGERS
        $('#bugButton').on('click', function(){
            $('.TopbarPageHeaderGlobalActions-omnibutton').trigger('mouseover').trigger('mousedown').trigger('click').trigger('keypress');
            setTimeout(function() {
                document.querySelector('.Omnibutton-addTask').click();
                setTimeout(function() {
                    $('.QuickAddTaskContents-descriptionEditor .ql-editor').html(bugXml);
                }, 500);
            }, 100);
        });

        // FEATURE TRIGGERS
        $('#featureButton').on('click', function(){
            $('.TopbarPageHeaderGlobalActions-omnibutton').trigger('mouseover').trigger('mousedown').trigger('click').trigger('keypress');
            setTimeout(function() {
                document.querySelector('.Omnibutton-addTask').click();
                setTimeout(function() {
                    $('.QuickAddTaskContents-descriptionEditor .ql-editor').html(featureXml);
                }, 500);
            }, 100);
        });

    });
});

// EVENT LOOPS
var target = document.querySelector('#asana_ui');
MutationObserver = window.MutationObserver || window.WebKitMutationObserver;

var observer = new MutationObserver(function(mutations, observer) {
    mutations.forEach(function(mutation) {

        if (mutation.addedNodes) {
            mutation.addedNodes.forEach(function(m){
                if (m.classList && (m.classList.contains('SingleTaskPane') || m.classList.contains('PotListPage-detailsPane'))){
                    timer = new Date().getTime();
                    updateCustomFields();
                }
            });
        }
    });
});

function updateCustomFields(){
    if (noiseButtonOn){
        var customFields = [];

        waitForEl('.CustomPropertyDetailsContainer-row', function(){
            $('.CustomPropertyDetailsContainer-row').each(function(i, item){
                customFields.push($(item).detach());
            });
            customFields.sort(function(a, b){
                var aName = $(a, 'label').text();
                var bName = $(b, 'label').text();
                return aName.localeCompare(bName);
            });
            for (var i = 0; i < customFields.length; i++){
                $('.CustomPropertyDetailsContainer-table').append(customFields[i]);
            }
        });
    }
}

observer.observe(target, {
    subtree: true,
    childList: true
});

setInterval(function() {
    // blackout loop
    if (blackoutButtonOn) {
        $('.RemovableAvatar-avatarRemoveButton').each(function(index,item){
            var followStyle = $(item).siblings('.Avatar').attr('style');
            if(blackoutProfileStyle.startsWith(followStyle.substring(0, followStyle.length - 13))) {
                $('.ToggleFollowButton-toggleText').trigger('click');
            }
        });
    }

    // Noise hiding loop
    if (noiseButtonOn) {
        $('.TaskStoryFeed-miniStory, .TaskList .MiniHeartButton, .TaskStoryFeed-separator').not('.StoryFeed-topSeparator, .swusana-noise-hidden, .swusana-always-ignore').each(function(index,item){
            if ($(item).text().indexOf(' created ') !== -1 || $(item).text().indexOf(' duplicated ') !== -1) {
                $(item).addClass('swusana-always-ignore');
            } else {
                $(item).addClass('swusana-noise');
                $(item).addClass('swusana-noise-hidden');
                $(item).hide();
            }
        });
    }

    if (timerButtonOn){
        var duration = parseInt((new Date().getTime() - timer)/1000);
        var date = new Date(null);
        date.setSeconds(duration); // specify value for SECONDS here
        var timeString = date.toISOString().substr(15, 4);
        $('#swusana-timer').html(timeString).css('background-color', duration < 30 ? '#b1f5b1' : duration < 60 ? '#eff953' : '#ff7e7e');
    }
}, 250);


function addGlobalStyle(css) {
    var head, style;
    head = document.getElementsByTagName('head')[0];
    if (!head) { return; }
    style = document.createElement('style');
    style.type = 'text/css';
    style.innerHTML = css;
    head.appendChild(style);
}

var waitForEl = function(selector, callback) {
  if (jQuery(selector).length) {
    callback();
  } else {
    setTimeout(function() {
      waitForEl(selector, callback);
    }, 100);
  }
};

// STYLE
var css =
    '.lunaui-grid-center-pane-container#center_pane_container { max-width: 100% !important}' +
    '.swusana-button { '+
        'margin-left: 8px;' +
        'margin-bottom: -4px;' +
        'opacity: 0.7' +
    '}' +
    '.swusana-button:hover { '+
        'opacity: 1;' +
    '}' +
    '.StoryFeed-miniStory+.StoryFeed-blockStory { margin-top: 0 !important; }' +
    '.swusana-button-on { '+
        'border-bottom: 3px solid #AAAAAA;' +
        'opacity: 1' +
    '}' +
    '.swusana-timer { margin-left: 6px;'+
      'background-color: #b1f5b1;'+
      'padding: 5px;'+
      'border-radius: 8px;'+
      'width: 28px;'+
      'border: 2px solid #555555;'+
    '}';
addGlobalStyle(css);
